### PR TITLE
Runnable pipeline on test data

### PR DESCRIPTION
**Pull request by mumichae**
Create a test dataset for development.
Testing can be done directly in the repository if `config/test_data.yaml` is used without modification. This PR doesn't address testing in pytest (`tests/`) or setting up a CI, which will be addressed in different PRs.

### CODE REVIEW

File: envs/env_vars_activate.sh
Comment (LuckyMD) [inline envs/env_vars_activate.sh:16]: missing an `_OLD` here

File: configs/test_data.yaml
Comment (LuckyMD) [inline configs/test_data.yaml:13]: isn't scIB-python meant to be the conv_env?

File: Snakefile
Comment (LuckyMD) [inline Snakefile:8]: what does the method statement mean here? Is this a regexp about the first letter not being a `/`?

File: README.md
Comment (LuckyMD) [inline README.md:39]: missing an `_OLD` here.

File: README.md
Comment (LuckyMD) [inline README.md:47]: you have this in the `activate_env.sh` and `deactivate_env.sh` files already... why not just refer to those faileds that they should be moved to the relevant directories?

File: ScibConfig.py
Comment (LuckyMD) [inline ScibConfig.py:89]: this is really cool... didn't know these operators were overloaded for Path variables.

File: Snakefile
Comment (LuckyMD) [inline Snakefile:12]: was this something that is just not used anymore? I.e. a separate cc scoring rule you used for testing?

File: data/generate_data.py
Comment (LuckyMD) [inline data/generate_data.py:12]: can just string together `astype(str).astype("category")`, no?

Comment (LuckyMD) [reply data/generate_data.py:12]: or just use category directly?

File: data/generate_data.py
Comment (LuckyMD) [inline data/generate_data.py:46]: why does category order matter?

File: data/generate_data.py
Comment (LuckyMD) [inline data/generate_data.py:68]: never call the other function? Why not give options?

**Review by LuckyMD — CHANGES_REQUESTED** [review]
A few typos and a few clarifications. @danielStrobl can you check if you can get the test pipeline running in server environment?

File: README.md
Comment (mumichae) [reply README.md:47]: These scripts didn't work for me, as `gsl-config` doesn't seem to be installed on my system. However only `LD` needed to be set in order to get `rpy2` running correctly. Also, Daniel mentioned something about `r-essentials` solving some of the issues

File: Snakefile
Comment (mumichae) [reply Snakefile:8]: Yes, I had issues with `snakemake` not finding `method=R/seurat`

File: Snakefile
Comment (mumichae) [reply Snakefile:12]: We decided to remove cc_score completely, as we didn't end up needing it

File: Snakefile
Comment (LuckyMD) [reply Snakefile:8]: I guess that's an issue with the config file then, no? Not really sth we have hardcoded like this somehwere?

File: README.md
Comment (LuckyMD) [inline README.md:66]: just a reminder here :)

File: configs/test_iso_AWS.yaml
Comment (LuckyMD) [inline configs/test_iso_AWS.yaml:15]: Could we move DATA_SCENARIOS to the bottom please? Looks a bit better there imho.

**Review by LuckyMD — CHANGES_REQUESTED** [review]
Some small beautifications, but this is looking really good :).

File: Snakefile
Comment (mumichae) [reply Snakefile:8]: We save R method output in a separate `R/` directory, which, when given the path, leads to ambiguity whether to use `method=seurat` from the R integration & conversion rules or `methd=R/seurat` from the python integration rule

Comment (LuckyMD) [reply Snakefile:8]: Ah, I see. Okay, thanks!

File: README.md
Comment (mumichae) [reply README.md:66]: I'd opt to move these TODOs to a new PR and focus on other issues first

File: data/generate_data.py
Comment (mumichae) [reply data/generate_data.py:20]: If I don't convert to str, conversion into R via `rpy2` will throw an error. But I agree, a one-liner is nicer

File: data/generate_data.py
Comment (mumichae) [reply data/generate_data.py:58]: To properly combine the cluster labels of both datasets, I guess. I copied the code from the [scanpy tutorial](https://scanpy-tutorials.readthedocs.io/en/latest/integrating-data-using-ingest.html) and will add the link in the docstring

File: data/generate_data.py
Comment (mumichae) [reply data/generate_data.py:80]: Yes, might want to consider that. However, the PBMC dataset did not make it through the pipeline, as it doesn't have enough gene features. I kept the code, in case we might want to use it, but would also consider removing it, if we don't need it

File: README.md
Comment (mumichae) [reply README.md:39]: 957bc20

File: envs/env_vars_activate.sh
Comment (mumichae) [reply envs/env_vars_activate.sh:16]: 957bc20

File: configs/test_iso_AWS.yaml
Comment (mumichae) [reply configs/test_iso_AWS.yaml]: 6640afb

File: data/generate_data.py
Comment (mumichae) [reply data/generate_data.py:63]: 375d957

File: data/generate_data.py
Comment (mumichae) [reply data/generate_data.py]: 375d957

File: README.md
Comment (mumichae) [reply README.md:47]: It might also be nicer to have all R-specific variables in the `${CONDA_PREFIX}/etc/conda/activate.d/activate-r-base.sh` and `${CONDA_PREFIX}/etc/conda/deactivate.d/deactivate-r-base.sh`, respectively

**Review by LuckyMD — COMMENTED** [review]
This looks pretty good. Did you just merge the new scIB-python env changes from master here?

File: data/generate_data.py
Comment (LuckyMD) [reply data/generate_data.py:85]: sounds good.

File: README.md
Comment (LuckyMD) [reply README.md:47]: Hmm... I think we can probably stick with env variables for now. If this doesn't work for you, could you just adapt the README to say there are 2 options to get this to work?

**Comment by LuckyMD** [comment]
@danielStrobl could you please test this so that we can merge?

**Comment by danielStrobl** [comment]
Hey! I've tested the pipeline with the new envs and I'm still getting some errors:
Metrics scanorama:
```
ValueError: batch "3" of batch_key "batch" has unequal number of entries before and after integration.before: 564 after: 550
```
Metrics with precomputed embeddings:
```
  File "scripts/metrics/metrics.py", line 249, in <module>
    trajectory_=trajectory_
  File "/home/icb/daniel.strobl/conda_scib/envs/scIB-python-21/lib/python3.7/site-packages/scIB/metrics.py", line 1855, in metrics
    cluster=False, n=n_isolated, verbose=False) if silhouette_ else np.nan
  File "/home/icb/daniel.strobl/conda_scib/envs/scIB-python-21/lib/python3.7/site-packages/scIB/metrics.py", line 325, in isolated_labels
    score = score_isolated_label(adata, label_key, cluster_key, label, cluster=cluster, verbose=verbose)
  File "/home/icb/daniel.strobl/conda_scib/envs/scIB-python-21/lib/python3.7/site-packages/scIB/metrics.py", line 391, in score_isolated_label
    score = silhouette(adata_tmp, group_key='group', **kwargs)
  File "/home/icb/daniel.strobl/conda_scib/envs/scIB-python-21/lib/python3.7/site-packages/scIB/metrics.py", line 50, in silhouette
    raise KeyError(f'{embed} not in obsm')
KeyError: 'X_pca not in obsm'
```
The integration part seems to work fine

**Comment by mumichae** [comment]
Thanks for testing the setup.

I'm getting the same error above with scanorama and mnn. This happens only when I add a batch effect to the counts in the `paul_15` dataset. Haven't figured out why this happens.

For the embedding error, use the `isolated_label_ASW_embed` branch of `scib`. You can find the PR here: https://github.com/theislab/scib/pull/213

**Comment by mumichae** [comment]
I identified the cause of errors in `mnnpy` and `scanorama`, see the [PR in scib](https://github.com/theislab/scib/pull/215).
TL;DR `Anndata.concatenate` renames annotations of the `batch_key`, if that key exists. In our test data, the batch key is `batch`, which is also the default for `batch_key` in `Anndata.concatenate`

**Comment by mumichae** [comment]
@danielStrobl Feel free to test the pipeline with the https://github.com/theislab/scib/tree/fix_concatenate for integration and `master` for the metrics. You can use the rule `integration` with the former branch until all methods run through, then switch to master and run the rest of the pipeline.

```commandline
snakemake --config-file config/test_data.yaml -j10 integration
```

**Comment by LuckyMD** [comment]
The fix_concatenate PR is good enough to be merged after a small confirmation on your part. Then this will be easier to test here again as well :). @danielStrobl it would be great if you could retest this as soon as @mumichae gives the go ahead with the other fix!

**Comment by mumichae** [comment]
scib PR is merged. I pushed some bugfixes, so don't forget to update @danielStrobl

**Review by LuckyMD — DISMISSED** [review]
Approved if this runs for @danielStrobl

**Comment by danielStrobl** [comment]
I've tested with the new envs and the updates now, everything is running through without any error

**Review by danielStrobl — APPROVED** [review]

**Comment by mumichae** [comment]
@LuckyMD Is this ready to be merged? Solving merge conflicts has revoked the approval. Also, I have adapted the documentation for environment variables, including the `QT_QPA_PLATFORM='offscreen'` variable as mentioned [here](https://github.com/theislab/scib/issues/209)

**Review by LuckyMD — APPROVED** [review]
in general this is good... just 1 comment on the README

File: README.md
Comment (mumichae) [reply README.md:36]: Added some explanation on how and where to copy these files

File: README.md
Comment (mumichae) [reply README.md:98]: Moved back to using the env files only and added more info on them in the README

**Comment by mumichae** [comment]
Would be good to have this merged asap, so I can move on to the scanpy update etc. here

File: ScibConfig.py
Comment (LuckyMD) [inline ScibConfig.py:82]: I think we only pass a separate HVG file to Seurat, but not other R methods. Have you changed this to do so for all R methods?

File: envs/env_vars_deactivate.sh
Comment (LuckyMD) [inline envs/env_vars_deactivate.sh:12]: Missing a `$` here in front of `LD_LIBRARY_PATH_OLD`

File: sanity_check/metrics_compare.R
Comment (LuckyMD) [inline sanity_check/metrics_compare.R:67]: This file should probably not be added here

File: scripts/integration/Snakefile
Comment (LuckyMD) [inline scripts/integration/Snakefile:23]: you don't need an input argument here?

File: scripts/metrics/metrics.py
Comment (LuckyMD) [inline scripts/metrics/metrics.py:139]: this is from iso label fix? I guess you merged master into here?

**Review by LuckyMD — CHANGES_REQUESTED** [review]
At least 1 typo (missing $) and a few questions on how you now store hvg files and some missing input args. Also 1 file added that doesn't belong in this PR (visualization of iso label fix results). Looks very close though.

File: sanity_check/metrics_compare.R
Comment (mumichae) [reply sanity_check/metrics_compare.R:67]: I needed it to compare different dependencies for the pipeline. Would you prefer it in the reproducibility repo instead?

File: scripts/integration/Snakefile
Comment (mumichae) [reply scripts/integration/Snakefile:23]: no, Snakemake is smart :D 

This notation works when the only input is `wildcards`. I removed the input file pattern, as the python script doesn't require any separate HVG input (it's computed in the script).

File: scripts/metrics/metrics.py
Comment (mumichae) [reply scripts/metrics/metrics.py:139]: Yes, I could remove that to reduce runtime slightly

Comment (mumichae) [reply scripts/metrics/metrics.py:139]: It's not merged from scib master though, I added it here for testing

File: ScibConfig.py
Comment (mumichae) [reply ScibConfig.py:82]: In the `integration_run_r` rule (check `scripts/integration/Snakefile`) I always pass a HVG file to the script. I'm assuming that script uses the HVGs for all R methods in question

Comment (mumichae) [reply ScibConfig.py:82]: From what I can see, the count object is subset by HVGs for all R methods, if it is given as an option

Comment (LuckyMD) [reply ScibConfig.py:82]: Hah... you're right... the subsetting is done later for R methods... only Seurat passes the list of HVGs. This is fine... was worried for a sec ^^.

File: sanity_check/metrics_compare.R
Comment (LuckyMD) [reply sanity_check/metrics_compare.R:67]: Hmm... this is a script we are using for our checks and not something that we show or expect others to use, no?  Would you want to expose this to the public?

File: scripts/metrics/metrics.py
Comment (LuckyMD) [reply scripts/metrics/metrics.py:139]: This is part of the iso label fix though, no? I mean... I think it's correct to do this. But not initially the remit of this PR. So just wondering if you also have this in scIB at the moment...

File: scripts/integration/Snakefile
Comment (LuckyMD) [reply scripts/integration/Snakefile:23]: cool :). Okay!

File: scripts/metrics/metrics.py
Comment (mumichae) [reply scripts/metrics/metrics.py:139]: I just double-checked and the results don't seem to change if I precompute PCA. Will remove it then

Comment (LuckyMD) [reply scripts/metrics/metrics.py]: this is what caused us to miss the iso label issue (as there was a PCA fallback). So I would keep this in here. I just wanted to make sure whether this code was already in scIB as well (before file moving).

**Review by LuckyMD — APPROVED** [review]
All good, just please add the precompute PCA false again (and check this is consistent across repos), then we can merge.

File: scripts/metrics/metrics.py
Comment (mumichae) [reply scripts/metrics/metrics.py]: The variable is not set to False for embed input in scib master

Comment (LuckyMD) [reply scripts/metrics/metrics.py]: okay, then nvm. mergeable :)

**Pull request merged**
